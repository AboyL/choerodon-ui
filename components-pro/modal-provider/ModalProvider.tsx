import React, {
  createContext,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
} from 'react';
import noop from 'lodash/noop';
import ModalContainer, { ModalContainerProps } from '../modal-container/ModalContainer';
import Modal, { ModalProps } from '../modal/Modal';

export interface ModalProviderProps extends ModalContainerProps {
  children?: ReactNode;
}

const ModalContext = createContext({ open: noop });

const ModalProvider = (props: ModalProviderProps) => {
  const { location, children } = props;
  const ref = useRef<ModalContainer>(null);
  const prepareToOpen = useMemo<(ModalProps & { children })[]>(
    () => [] as (ModalProps & { children })[],
    [],
  );

  const open = useCallback((modalProps: ModalProps & { children }) => {
    const container = ref.current;
    if (container) {
      const close = async (destroy?: boolean) => {
        const { onClose = noop } = modalProps;
        if ((await onClose()) !== false) {
          if (destroy) {
            container.close({ ...modalProps, destroyOnClose: true });
          } else {
            container.close(modalProps);
          }
        }
      };

      const show = () => {
        container.top().open(modalProps);
      };

      const update = newProps => {
        container.top().update({ ...modalProps, ...newProps });
      };

      modalProps = {
        close,
        update,
        ...Modal.defaultProps,
        ...modalProps,
      };
      container.top().open(modalProps);

      return {
        close,
        open: show,
        update,
      };
    }
    prepareToOpen.push(modalProps);
  }, []);

  useEffect(() => {
    if (ref.current) {
      prepareToOpen.forEach(prepare => open(prepare));
    }
  }, [ref, open]);

  const context = { open };

  return (
    <ModalContext.Provider value={context}>
      <ModalContainer ref={ref} location={location} />
      {children}
    </ModalContext.Provider>
  );
};

const useModal = () => {
  return useContext(ModalContext);
};

const injectModal = Target => {
  const Hoc = props => {
    const modal = useModal();
    return <Target {...props} Modal={modal} />;
  };

  Hoc.displayName = `${Target.displayName || 'Anonymous'}-with-inject-modal`;

  return Hoc;
};

ModalProvider.displayName = 'ModalProvider';

ModalProvider.useModal = useModal;

ModalProvider.injectModal = injectModal;

export default ModalProvider;
