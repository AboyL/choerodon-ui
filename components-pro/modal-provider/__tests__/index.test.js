import React from 'react';
import { mount } from 'enzyme';
import ModalProvider from '..';

describe('ModalProvider', () => {
  it('renders ModalProvider correctly', () => {
    const wrapper = mount(<ModalProvider />);
    // expect(wrapper).toMatchSnapshot();
  });
});
