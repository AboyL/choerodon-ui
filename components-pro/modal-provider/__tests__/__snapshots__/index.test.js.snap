// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider renders ModalProvider correctly 1`] = `
<ModalProvider>
  <ModalContainer>
    <Portal
      containerInfo={
        <div
          class="c7n-pro-modal-container"
        >
          <div
            class="c7n-pro-mask c7n-pro-mask-wrapper"
            hidden=""
          />
        </div>
      }
    >
      <Animate
        animation={Object {}}
        component=""
        componentProps={Object {}}
        hiddenProp="hidden"
        onEnd={[Function]}
        transitionAppear={true}
        transitionEnter={true}
        transitionLeave={true}
        transitionName="fade"
      >
        <AnimateChild
          animation={Object {}}
          key="animate_1582707529919"
          transitionAppear={true}
          transitionEnter={true}
          transitionLeave={true}
          transitionName="fade"
        >
          <Mask
            hidden={true}
            key="animate_1582707529919"
            onClick={[Function]}
            onMouseDown={[Function]}
            style={Object {}}
            suffixCls="mask"
          >
            <div
              className="c7n-pro-mask c7n-pro-mask-wrapper"
              hidden={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              style={Object {}}
            />
          </Mask>
        </AnimateChild>
      </Animate>
    </Portal>
  </ModalContainer>
</ModalProvider>
`;
